core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java

```diff
@@ -26,42 +26,33 @@
 import org.elasticsearch.indices.query.IndicesQueriesRegistry;
 
 import java.io.IOException;
+import java.util.Objects;
 
 public class QueryParseContext {
 
     private static final ParseField CACHE = new ParseField("_cache").withAllDeprecated("Elasticsearch makes its own caching decisions");
     private static final ParseField CACHE_KEY = new ParseField("_cache_key").withAllDeprecated("Filters are always used as cache keys");
 
-    private XContentParser parser;
-    private ParseFieldMatcher parseFieldMatcher = ParseFieldMatcher.EMPTY;
+    private final XContentParser parser;
 
     private IndicesQueriesRegistry indicesQueriesRegistry;
 
-    public QueryParseContext(IndicesQueriesRegistry registry) {
-        this.indicesQueriesRegistry = registry;
+    public QueryParseContext(IndicesQueriesRegistry registry, XContentParser parser, ParseFieldMatcher parseFieldMatcher) {
```

Comment:
I added this mainly because in tests we sometimes use ParseFields.EMPTY, sometimes ParseFields.STRICT. I think it would generally be much better to only keep the ParseFieldMatcher in one place, e.g. the parser itself, so I will look if I can get rid of this.